### PR TITLE
[Macros] Implement unqualified lookup for global macro-expanded peer declarations.

### PR DESCRIPTION
**Pull request by hborla**
This change adds `MacroDecl::getIntroducedNames`, which populates the given vector with a list of names introduced by the macro decl. This list of macro-covered names is used in unqualified lookup to lazily expand declarations with attached macros covering a requested name. This can also be used later for name validation of attached macros.

Resolves: rdar://104345950

### CODE REVIEW

**Comment by hborla** [comment]
@swift-ci please smoke test

**Comment by hborla** [comment]
@swift-ci please smoke test

File: lib/AST/Module.cpp [hunk 1]
```diff
@@ -309,6 +325,53 @@ void SourceLookupCache::addToMemberCache(Range decls) {
...
+    // lookup for macro arguments, which will recursively populate the
+    // auxiliary decl cache and cause request cycles.
+    //
+    // We do not need a fully resolved macro until expansion. Instead, we
+    // conservatively consider peer names for all macro declarations with a
+    // custom attribute name. Unqualified lookup for that name will later
+    // invoke expansion of the macro, and will yield no results if the resolved
+    // macro does not produce the requested name, so the only impact is possibly
+    // expanding earlier than needed / unnecessarily looking in the top-level
+    // auxiliary decl cache.
```
Comment (DougGregor) [inline lib/AST/Module.cpp:344]: I like this approach a lot!

**Review by DougGregor — APPROVED** [review]
This is looking great!

File: lib/AST/Decl.cpp [hunk 2]
```diff
@@ -10027,7 +10049,68 @@ const MacroRoleAttr *MacroDecl::getMacroRoleAttr(MacroRole role) const {
...
+
+    case MacroIntroducedDeclNameKind::Prefixed: {
+      if (!attachedTo)
+        break;
+
+      auto baseName = attachedTo->getBaseName();
+      std::string prefixedName;
+      llvm::raw_string_ostream out(prefixedName);
+      out << expandedName.getIdentifier();
+      out << baseName.getIdentifier();
```
[the remark this answers is not in the document]
Comment (hborla) [reply lib/AST/Decl.cpp]: Good to know, fixed!

File: lib/AST/Decl.cpp [hunk 3]
```diff
@@ -10027,7 +10049,68 @@ const MacroRoleAttr *MacroDecl::getMacroRoleAttr(MacroRole role) const {
...
+      out << expandedName.getIdentifier();
+
+      Identifier nameId = ctx.getIdentifier(suffixedName);
+      names.push_back(DeclName(nameId));
+      break;
+    }
+
+    case MacroIntroducedDeclNameKind::Arbitrary:
+      // Only freestanding and member macros can have arbitrary names.
+      assert(role == MacroRole::Member || role == MacroRole::Declaration);
```
[the remark this answers is not in the document]
Comment (hborla) [reply lib/AST/Decl.cpp]: I replaced the assertion with a `FIXME` for now, and I'm working on adding support for `arbitrary` through this mechanism locally.

**Comment by hborla** [comment]
@swift-ci please smoke test

**Pull request merged**